File: py/1081.smallest-subsequence-of-distinct-characters.py
#
# @lc app=leetcode id=1081 lang=python3
#
# [1081] Smallest Subsequence of Distinct Characters
#

# @lc code=start
# t:O(N), s:O(1)
class Solution:
    def smallestSubsequence(self, s: str) -> str:
        last_seen = [-1]*26
        seen = [False]*26
        for idx, val in enumerate(s):
            last_seen[ord(val)-ord('a')] = idx
        res = []
        for idx, val in enumerate(s):
            if not seen[ord(val)-ord('a')]:
                while len(res) and val < res[-1] and last_seen[ord(res[-1])-ord('a')] > idx:
                    seen[ord(res[-1])-ord('a')] = False
                    res.pop()
                res.append(val)
                seen[ord(val)-ord('a')] = True
        return ''.join(res)

# @lc code=end

File: cpp/1081.smallest-subsequence-of-distinct-characters.cpp
/*
 * @lc app=leetcode id=1081 lang=cpp
 *
 * [1081] Smallest Subsequence of Distinct Characters
 */
#include <string>
#include <vector>
using namespace std;
// @lc code=start
// t:O(N), s:O(1)
class Solution
{
public:
    string smallestSubsequence(string s)
    {
        vector<int> last_seen(26, -1);
        int w = s.size();
        for (int i = 0; i != w; ++i)
        {
            last_seen[s[i] - 'a'] = i;
        }
        string res = "";
        vector<bool> seen(26);
        for (int i = 0; i != w; ++i)
        {
            char c = s[i];
            if (!seen[c - 'a'])
            {
                while (!res.empty() && c < res.back() && last_seen[res.back() - 'a'] > i)
                {
                    seen[res.back() - 'a'] = false;
                    res.pop_back();
                }
                res.push_back(c);
                seen[c - 'a'] = true;
            }
        }
        return res;
    }
};
// @lc code=end
